lib/rucio/core/did.py

```diff
@@ -1206,6 +1207,8 @@ def delete_dids(dids, account, expire_rules=False, session=None, logger=logging.
             collection_replica_clause.append(and_(models.CollectionReplica.scope == did['scope'],
                                                   models.CollectionReplica.name == did['name']))
             did_followed_clause.append(and_(models.DidsFollowed.scope == did['scope'], models.DidsFollowed.name == did['name']))
+            if did['did_type'] == DIDType.DATASET:
+                dataset_clause.append(and_(models.DataIdentifier.scope == did['scope'], models.DataIdentifier.name == did['name']))
```

Comment:
```
dataset_clause.append(and_(models.DataIdentifierAssociation.scope == did['scope'], models.DataIdentifierAssociation.name == did['name']))
```
otherwise it will create an implicit cartesian product and not do what you think it should do